File: Data Structures & Algorithms/course-schedule/submission-1.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        premap = {i:[] for i in range(numCourses)}
        visited = set()
        for cur,pre in prerequisites:
            premap[cur].append(pre)
        def dfs(cur):
            if cur in visited:
                return False
            if len(premap[cur])==0:
                return True
            visited.add(cur)
            for c in premap[cur]:
                if not dfs(c):
                    return False
            #回到上一层
            premap[cur].clear()
            visited.remove(cur)
            return True
        for c in range(numCourses):
            if not dfs(c):
                return False
        return True

File: Data Structures & Algorithms/count-paths/submission-1.cpp
class Solution {
public:
    int dfs(vector<vector<int>> &cache,int i,int j,int m, int n) {
        if (i>=m || j>=n) return 0;
        if (i==m-1 && j==n-1) return 1;
        if (cache[i][j]) return cache[i][j];
        return cache[i][j] = dfs(cache,i+1,j,m,n) + dfs(cache,i,j+1,m,n);
    }
    int uniquePaths(int m, int n) {
        vector<vector<int>> cache(m,vector<int>(n,0));
        cache[m-1][n-1] = 1;
        return dfs(cache,0,0,m,n);
    }
};
